dashboard/test/models/user_test.rb

```diff
@@ -1758,6 +1758,41 @@ def complete_script_for_user(user, script, completed_date = Time.now)
     assert_equal User.hash_email('first@email.com'), student.primary_contact_info.hashed_email
   end
 
+  test 'update_primary_contact_info fails safely if the new email is already taken for sponsored user' do
+    taken_email = 'taken@example.org'
+    create :student, email: taken_email
+    update_primary_contact_info_fails_safely_for \
+      create(:student_in_picture_section, :multi_auth_migrated),
+      user: {email: taken_email}
+  end
+
+  test 'update_primary_contact_info fails safely if the new email is already taken for email user' do
+    taken_email = 'taken@example.org'
+    create :student, email: taken_email
+    update_primary_contact_info_fails_safely_for \
+      create(:student, :with_migrated_email_authentication_option),
+      user: {email: taken_email}
+  end
+
+  test 'update_primary_contact_info fails safely if the new email is already taken for oauth user' do
+    taken_email = 'taken@example.org'
+    create :student, email: taken_email
+    update_primary_contact_info_fails_safely_for \
+      create(:student, :with_migrated_google_authentication_option),
+      user: {email: taken_email}
+  end
+
+  def update_primary_contact_info_fails_safely_for(user, *params)
+    original_primary_contact_info = user.primary_contact_info
+
+    refute_creates_or_destroys AuthenticationOption do
```

Comment:
you should be able to use [assert_no_difference](https://apidock.com/rails/ActiveSupport/Testing/Assertions/assert_no_difference) here rather than aliasing assert_does_not_create

Reply:
nope, although does assert_does_not_create check that nothing has been deleted as well? i thought it only checked for creations (i can't find it in the docs either)

Reply:
It's not in the docs because it's our own code, and using `assert_no_difference` so it actually checks for both create and destroy:

https://github.com/code-dot-org/code-dot-org/blob/d995e31289c6fb0753b4bc0f2c3c6f14f75ca751/dashboard/test/test_helper.rb#L132-L136